lib/Frontend/Frontend.cpp

```diff
@@ -261,236 +264,278 @@ ModuleDecl *CompilerInstance::getMainModule() {
+  const StringRef implicitHeaderPath =
+      Invocation.getFrontendOptions().ImplicitObjCHeaderPath;
+  auto clangImporter =
+      static_cast<ClangImporter *>(Context->getClangModuleLoader());
+  if (implicitHeaderPath.empty() ||
+      clangImporter->importBridgingHeader(implicitHeaderPath, MainModule))
+    return nullptr;
+  ModuleDecl *importedHeaderModule = clangImporter->getImportedHeaderModule();
+  assert(importedHeaderModule);
+  return importedHeaderModule;
+}
+
+void CompilerInstance::getImplicitlyImportedModules(
+    SmallVectorImpl<ModuleDecl *> &importModules) {
+  SharedTimer timer("performSema-getImplicitlyImportedModules");
+  for (auto &ImplicitImportModuleName :
+       Invocation.getFrontendOptions().ImplicitImportModuleNames) {
+    if (Lexer::isIdentifier(ImplicitImportModuleName)) {
+      auto moduleID = Context->getIdentifier(ImplicitImportModuleName);
+      ModuleDecl *importModule =
+          Context->getModule(std::make_pair(moduleID, SourceLoc()));
+      if (importModule) {
+        importModules.push_back(importModule);
+      } else {
+        Diagnostics.diagnose(SourceLoc(), diag::sema_no_import,
+                             ImplicitImportModuleName);
+        if (Invocation.getSearchPathOptions().SDKPath.empty() &&
+            llvm::Triple(llvm::sys::getProcessTriple()).isMacOSX()) {
+          Diagnostics.diagnose(SourceLoc(), diag::sema_no_import_no_sdk);
+          Diagnostics.diagnose(SourceLoc(), diag::sema_no_import_no_sdk_xcrun);
+        }
+      }
+    } else {
+      Diagnostics.diagnose(SourceLoc(), diag::error_bad_module_name,
+                           ImplicitImportModuleName, false);
+    }
   }
+}
 
-  bool hadLoadError = false;
+void CompilerInstance::createREPLFile(ImplicitImports &implicitImports) const {
+  SharedTimer timer("performSema-createREPLFile");
+  auto *SingleInputFile = new (*Context) SourceFile(
+      *MainModule, Invocation.getSourceFileKind(), None, implicitImports.kind,
+      Invocation.getLangOptions().KeepTokensInSourceFile);
+  MainModule->addFile(*SingleInputFile);
+  addAdditionalInitialImportsTo(SingleInputFile, implicitImports);
+}
 
-  // Parse all the partial modules first.
-  for (auto &PM : PartialModules) {
-    assert(PM.ModuleBuffer);
-    if (!SML->loadAST(*MainModule, SourceLoc(), std::move(PM.ModuleBuffer),
-                      std::move(PM.ModuleDocBuffer)))
-      hadLoadError = true;
-  }
+std::unique_ptr<DelayedParsingCallbacks>
+CompilerInstance::computeDelayedParsingCallback() {
+  DelayedParsingCallbacks *cb = nullptr;
+  if (Invocation.isCodeCompletion())
+    cb = new CodeCompleteDelayedCallbacks(SourceMgr.getCodeCompletionLoc());
+  else if (Invocation.isDelayedFunctionBodyParsing())
+    cb = new AlwaysDelayedCallbacks;
 
-  // Then parse all the library files.
-  for (auto BufferID : BufferIDs) {
-    if (BufferID == MainBufferID)
-      continue;
+  std::unique_ptr<DelayedParsingCallbacks> up(cb);
+  return up;
```

Comment:
Nitpick: you should be able to do this all in one line, as

`return std::unique_ptr<DelayedParsingCallbacks>(cb);`

…and, if you so choose, you could collapse it even further:

```c++
if (Invocation.isCodeCompletion()) {
  return llvm::make_unique<CodeCompleteDelayedCallbacks>(
      SourceMgr.getCodeCompletionLoc());
}
if (Invocation.isDelayedFunctionBodyParsing())
  return llvm::makeUnique<AlwaysDelayedCallbacks>();
return nullptr;
```

Is this better? Well, it means there's never a dangling `new` that could accidentally leak. I don't think it's critical in this case, but `make_unique` is a nice tool to have in your toolbox, at least. (It's `std::make_unique` in C++14, but we haven't made that a requirement yet.)

Reply:
Love it! Thanks for make_unique.